### PR TITLE
#5 Refactor browse match and bug fix

### LINKED ISSUE #216: Change to challenge card options being sent to handlebars
When writing the filter function for organisations i couldn't reuse a function that was created for filtering challenges:

specifically this reduce function:

```
     el.tags.reduce((bool, tag) =>
        tag[0].id === filtered[0] && tag[1].id === filtered[1] ? true : bool
      , false)
    );
```

The problem being caused by this reusable function: 
```
function challengesFromIds (challenges, ids) {
  return ids.map(id => {
    var challengeCard = JSON.parse(challenges[id]);
    var tagData = getTagNames(challengeCard.tags);
    return Object.assign({}, challengeCard, {tagData});
  });
}
```

As it overwrites the `challenge.tags` key from `[ [0,0], [1,1] ]` to `[ {id:id, name:name}, {id:id, name:name} ]`

But the `organisation.tags`is still of the form `[ [0,0], [1,1] ]`.

To get around the problem, rather than overwriting the `.tags` key, i made a new key called `.tagsData` which holds the name and ids, but the objects still have the `.tags` key which only holds the `tag_id`s

Here is an example challenge card before it gets handed over to handlebars:
```
 id: 3,
  date: 1475220201284,
  org_id: 1,
  creator_id: 3,
  title: 'Challenge Number 4',
  description: 'Who should I...?',
  active: true,
  tags: [ [ 1, 1 ], [ 1, 3 ], [ 5, 1 ], [ 6, 2 ] ],
  tagData:
   [ { id: [ 1, 1 ], name: 'AUTOMOTIVE AND TRANSPORT MANUFACTURING' },
     { id: [ 1, 3 ], name: 'CHEMICALS' },
     { id: [ 5, 1 ], name: 'SECONDARY EDUCATION' },
     { id: [ 6, 2 ], name: 'DESIGN FOR DISASSEMBLY' } ],
  shared_by: 'dwyl' }
```

I wanted to let you know of the change @mk4111 as it affects `challenge.html` `browse.html` `org-details-view.js` and `browse-challenges-view.js` and you can review it in the next PR that I make. Let me know if you have any thoughts/concerns.

I will take a look into other files to see if any other changes should be made to keep the naming consistent across all codebase.

### LINKED ISSUE #218: admin cannot see deactivated orgs in browse view
Admin should be able to see active and inactive orgs in browse view.

@asimonw 2 questions for admin viewing challenges:

1. Should they see both active and inactive challenges?
2. Should they see challenges created by an inactive org?

### LINKED ISSUE #220: bugs and todos
Will flesh this issue out but I have noted down some todos as a result of the work i did this morning:

Some things to do that I will go onto address in the next PR:

- [x] add tests
- [x] check tagsData change whether it should be added elsewhere
- [x] put reusable browse functions in helpers
- [x] remove active and contact from primary user contact view

~~- [ ] check outstanding todos in the project~~
~~- [ ] add filter to browse orgs, click organisations, filter disappears~~ --> see #226
~~- [ ] add query validation on new browse routes --> see #225~~

### LINKED ISSUE #227: contact list bug
Primary users should not be able to see user type and activation status on their contact list.

Contact list needs amending to show:
```
Name
job_title at Organisation_name
```

### PR DESCRIPTION
**Pull request by JMurphyWeb**
~~**Note base needs changing to master and @nelsonic needs assigning when @mk4111 has reviewed.**~~

Resolves #216 
Resolves #218 
Resolves #220
Resolves #227

### CODE REVIEW

File: server/handlers/browse-view.js [hunk 1]
```diff
@@ -18,29 +18,32 @@ module.exports = (request, reply) => {
     Hoek.assert(!error, 'redis error');
     request.redis.LRANGE('challenges', 0, -1, (error, allChallengesString) => {
       Hoek.assert(!error, 'redis error');
-      // Filter out inactive orgs
-      var activeOrgs = helpers.filterActive(allOrgsString);
+      var orgs = helpers.parseArray(allOrgsString);
+      var challenges = helpers.parseArray(allChallengesString);
+
+      // Filter out unwanted orgs depending on permissions
+      var browsableOrgs = getBrowsable(loggedIn.scope, orgs);
```
Comment (mk4111) [inline server/handlers/browse-view.js]: _really_ like this!

File: server/handlers/helpers.js [hunk 2]
```diff
@@ -126,8 +116,23 @@ helpers.sortAlphabetically = (key) => (arr) =>
...
+
+// add the names to all tagIds that this function receives
+helpers.getTagNames = (tagIds) => {
+  var allTags = require('../../tags/tags.json');
+  return tagIds.map(helpers.getTagFromId(allTags));
+}
+
+// map through the inner function which takes a tag id, and returns named object for that tag
+helpers.getTagFromId = (allTags) => (id) =>
+  id && allTags[id[0]] && allTags[id[0]].tags[id[1]] && {
```
Comment (mk4111) [inline server/handlers/helpers.js]: @JMurphyWeb - is there a benefit of separating getTagFromId from getTagNames?

**Comment by mk4111** [comment]
Just chatted to @JMurphyWeb about this and discussed the above point.

Ready for @nelsonic to review (please)

File: server/handlers/helpers.js (hunk 2 above)
Comment (JMurphyWeb) [inline server/handlers/helpers.js]: getTagFromId is used in another file so we are going to keep it as is.

**Comment by nelsonic** [comment]
👀

File: server/handlers/browse-view.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  path: '/people',
+  name: 'People'
+}];
+
+module.exports = (request, reply) => {
+  var loggedIn = request.auth.credentials;
+  var permissions = helpers.getPermissions(loggedIn, 'scope', 'admin');
+  // get all orgs
+  request.redis.LRANGE('organisations', 0, -1, (error, allOrgsString) => {
+    Hoek.assert(!error, 'redis error');
```
Comment (nelsonic) [inline server/handlers/browse-view.js]: you _can_ use `reqest.handleError` here see: https://github.com/dwyl/hapi-error#handleerror-everywhere

**Comment by nelsonic** [comment]
@JMurphyWeb looks good. (_@mk4111 thanks for doing initial review._ 👍 )
merging. 🎉

**Pull request merged**